src/LogicTasks/Helpers.hs

```diff
@@ -55,6 +55,12 @@ example correct s = indent $ do
 reject :: OutputMonad m => State (Map Language String) () -> LangM m
 reject  = refuse . indent . translate
 
+applyForAll :: OutputMonad m => [a] -> (a -> LangM m) -> LangM m
+applyForAll [] _ = pure ()
+applyForAll (x:xs) m = do
+  m x
+  applyForAll xs m
+  pure ()
```

Comment:
I'm not quite familiar with `forM_`. Replacing `applyForAll` simply with `forM_` does not work.

Reply:
Does not work in what sense? Does it give a type error? Or does something wrong at runtime?

I would expect the functiion to be equivalent to `forM_` provided the line 63 would be omitted. And this line can be omitted, right? Since the recursion will hit `pure ()` anyway at the end, via the base case.

Reply:
It throws a type error.
```
• Could not deduce (Monad
                      (Control.Monad.Output.Generic.GenericLangM
                         output-monad-0.1.0.1:Control.Monad.Output.Report.Language m))
    arising from a use of ‘forM_’
  from the context: (OutputMonad m, MonadIO m)
    bound by the type signature for:
               completeGrade :: forall (m :: * -> *).
                                (OutputMonad m, MonadIO m) =>
                                FilePath -> LegalPropositionInst -> [Int] -> LangM m
    at E:\Dev\Uni\logic-tasks\src\LogicTasks\Syntax\IllegalFormulas.hs:75:1-99
• In the first argument of ‘($)’, namely
    ‘forM_ (correctTrees inst)’
  In a stmt of a 'do' block:
    _ <- forM_ (correctTrees inst)
           $ \ x
               -> do code (display x)
                     image $=<< liftIO $ cacheTree (transferToPicture x) path
                     return ()
  In the second argument of ‘($)’, namely
    ‘do example (show (toList (serialsOfWrong inst)))
          $ do english "A possible solution for this task is:"
               german "Eine mögliche Lösung für die Aufgabe ist:"
        instruct
          $ do english
                 "The following syntax trees represent the well-formed formulae:"
               german
                 "Die folgenden Syntaxbäume entsprechen den wohlgeformten Formeln:"
        forM_ (correctTrees inst) $ \ x -> do ...
        return ()’typecheck(-Wdeferred-type-errors)
```
This is the output if I replace `applyForAll` with `forM_` (in IllegalFormulas.hs). Removing `pure ()` does not fix it.

Reply:
Ah, okay, the short answer is that I should have suggested `Data.Foldable.for_` instead of `Control.Monad.forM_`, because that one only requires an `Applicative` context instead of `Monad`.

Reply:
The long answer is, or actually really a question for @marcellussiegburg, how it can be that `GenericLangM Language m` is not recognized as a `Monad`, even when explicitly adding a `Monad m` constraint, when on the other hand `instance MonadTrans (GenericLangM l)` exists in the `output-monad` package and `MonadTrans (GenericLangM l)` seems to indeed require that `forall m. Monad m => Monad (GenericLangM l m)`.

Reply:
This is an interesting question and maybe `MonadTrans` is indeed abused here. One could argue that none of the `MonadTrans` laws hold for the instance of `GenericLangM`. First of all, `MonadTrans` itself does not explicitly require a Monad instance of its outer class (i.e. it is defined as `class MonadTrans t` rather than `class Monad t => MonadTrans t`). Of course the question arises what does this make of the laws in this case:
1. `lift . return = return`: `return` is not defined, as `GenericLangM l  m` is no `Monad`, but `return` is just a synonym for `pure`. So it would be possible to argue that it sort of holds as `lift . pure = pure` does hold. But to say the least we can say that it does not defy the law as it is not possible to write `return` (but `lift . return` is).
2. `lift (m >>= f) = lift m >>= (lift . f)`: Again, `>>=` is not defined, argument passing using bind is not possible, i.e. it is not possible to write `lift m >>= (lift . f)`, so it might not be that the law holds; but it does not defy it either. The closest we could write is `lift (m *> f) = lift m *> lift f` which would hold.

The reasons why it exists might be weak:
- compatibility with older code
- convenience

But, as this does seem to raise questions. Maybe this should be changed? There could be a `FunctorTrans` class defined within `output-monad` having the same function `lift` but defining the above stated (modified) laws. In turn it would then be required to import this other `lift` when writing `output-monad` code. Confusing both would be no issue, as the first law behaves equivalent and the second law is different in the sense that `f` is of type `a -> m b` while in the `FunctorTrans` case of type `f b`, i.e. the compiler would throw errors if the "wrong" `lift` would be used. `GenericLangM` could would require `FunctorLift` because of its type. @jvoigtlaender What do you think?

And why do we require `lift` at all? In order to be able to have (restricted) `IO`-computations, that can be lifted into the inner monad (like reading or computing image files).

Reply:
> First of all, `MonadTrans` itself does not explicitly require a Monad instance of its outer class (i.e. it is defined as `class MonadTrans t` rather than `class Monad t => MonadTrans t`)

Is that so (and will it be for long)?

My statement yesterday:
> `MonadTrans (GenericLangM l)` seems to indeed require that `forall m. Monad m => Monad (GenericLangM l m)`

was based on this: https://hackage.haskell.org/package/transformers-0.6.1.1/docs/src/Control.Monad.Trans.Class.html#MonadTrans

So, it appears that at least from GHC 8.6 and `transformers-0.6` onwards one cannot have a `MonadTrans` if one doesn't have a `Monad` instance for the "outer type".

Maybe this forces our hands as to how to realize `lift` in `output-monad` (in the near future)?